### ISSUE TITLE
some way to define self types

### ISSUE THREAD

**Issue by graingert** [Shoobx/mypy-zope#44]
eg I might want to define

```python
class ICopyable(zope.interface.Interface):
    def copy(self: T) -> T:
        pass
```

but I get "Interface methods should not have 'self' argument"

**Comment by graingert** [comment]
I can use this, but then I lose the type of the function in `@implements`

```python
class ICopyable(zope.interface.Interface):
    @zope.interface.fromMethod
    def copy(self: T) -> T:
        pass
```

**Comment by kedder** [comment]
Right, interface method cannot have `self` argument - it won't work at runtime (my guess). However, we might make `@zope.interface.interface.fromMethod` trick work, if that's acceptable in your case. It feels a bit like a hack though. It isn't anywhere in the official docs.

The trick here is to make it work both at runtime and during static checking.
